test/unit/deprecate_warning_tests.js

```diff
@@ -0,0 +1,241 @@
+    metadata: { requires: { node: '>=6.0.0' } },
+    test: function(done) {
+      setupMultFunctionswithSameOptions();
+      process.nextTick(() => {
+        expect(messages).to.deep.equal([
+          'f1 option [maxScan]' + defaultMessage,
+          'f2 option [maxScan]' + defaultMessage
+        ]);
+        expect(messages).to.have.a.lengthOf(2);
+        done();
+      });
+    }
+  });
+
+  it('multiple functions with the same deprecated options should both warn', {
+    metadata: { requires: { node: '<6.0.0' } },
+    test: function(done) {
+      setupMultFunctionswithSameOptions();
+      ensureCalledWith(console.error, [
+        'f1 option [maxScan]' + defaultMessage,
+        'f2 option [maxScan]' + defaultMessage
+      ]);
+      expect(console.error).to.have.been.calledTwice;
+      done();
+    }
+  });
+
+  function setupNoOptions() {
+    const f = makeTestFunction({
+      name: 'f',
+      deprecatedOptions: deprecatedOptions,
+      optionsIndex: 0
+    });
+    f({});
+  }
+
+  it('should not warn if no deprecated options passed in', {
+    metadata: { requires: { node: '>=6.0.0' } },
+    test: function(done) {
+      setupNoOptions();
+      process.nextTick(() => {
+        expect(messages).to.have.a.lengthOf(0);
+        done();
+      });
+    }
+  });
+
+  it('should not warn if no deprecated options passed in', {
+    metadata: { requires: { node: '<6.0.0' } },
+    test: function(done) {
+      setupNoOptions();
+      expect(console.error).to.have.not.been.called;
+      done();
+    }
+  });
+
+  function setupUserMsgHandler() {
+    const customMsgHandler = (name, option) => {
+      return 'custom msg for function ' + name + ' and option ' + option;
+    };
+
+    const f = makeTestFunction({
+      name: 'f',
+      deprecatedOptions: deprecatedOptions,
+      optionsIndex: 0,
+      msgHandler: customMsgHandler
+    });
+
+    f({ maxScan: 5, snapshot: true, fields: 'hi' });
+  }
+
+  it('should use user-specified message handler', {
+    metadata: { requires: { node: '>=6.0.0' } },
+    test: function(done) {
+      setupUserMsgHandler();
+      process.nextTick(() => {
+        expect(messages).to.deep.equal([
+          'custom msg for function f and option maxScan',
+          'custom msg for function f and option snapshot',
+          'custom msg for function f and option fields'
+        ]);
+        expect(messages).to.have.a.lengthOf(3);
+        done();
+      });
+    }
+  });
+
+  it('should use user-specified message handler', {
+    metadata: { requires: { node: '<6.0.0' } },
+    test: function(done) {
+      setupUserMsgHandler();
+      ensureCalledWith(console.error, [
+        'custom msg for function f and option maxScan',
+        'custom msg for function f and option snapshot',
+        'custom msg for function f and option fields'
+      ]);
+      expect(console.error).to.have.been.calledThrice;
+      done();
+    }
+  });
+
+  function setupOncePerOption() {
```

Comment:
All of these functions that involve `setup` can be done in a nested `describe` block using `beforeEach`